qemu/tests/vsock_test.py

```diff
@@ -0,0 +1,154 @@
+    :return: the shell session with opened vsock listening process
+    """
+    nc_vsock_cmd = "%s -l %s" % (nc_vsock_bin, port)
+    session.read_nonblocking(0, timeout=10)
+    logging.info("Listening to the vsock port from guest: %s" % nc_vsock_cmd)
+    session.sendline(nc_vsock_cmd)
+
+
+def nc_vsock_connect(nc_vsock_bin, guest_cid, port):
+    """
+    Connect to vsock port from host, cmd: nc-vsock $guest_cid $port
+
+    :param nc_vsock_bin: path of binary nc-vsock
+    :param guest_cid: guest cid to connect
+    :param port: port to connect
+    :return: The vsock session from host side, being waiting for input
+    """
+    nc_vsock_cmd = "%s %s %s" % (nc_vsock_bin, guest_cid, port)
+    logging.info("Connect to the vsock port on host: %s" % nc_vsock_cmd)
+    return aexpect.Expect(
+        nc_vsock_cmd,
+        auto_close=False,
+        output_func=utils_misc.log_line,
+        output_params=("vsock_%s_%s" % (guest_cid, port),))
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Vsock basic function test
+
+    1. Boot guest with vhost-vsock-pci device
+    2. Download and compile nc-vsock on both guest and host
+    3. Start listening inside guest, nc-vsock -l $port
+    4. Connect guest CID from host, nc-vsock $guest_cid $port
+    5. Input character, e.g. 'Hello world'
+    6. Check if guest receive the content correctly
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+
+    def clean(tmp_file=None):
+        """ Clean the environment """
+        cmd_rm = "rm -rf %s*" % nc_vsock_bin
+        if tmp_file:
+            cmd_rm += "; rm -rf %s" % tmp_file
+        session.cmd_output_safe(cmd_rm)
+        process.system(cmd_rm, ignore_status=True)
+        if host_vsock_session.is_alive():
+            host_vsock_session.close()
+        session.close()
+
+    vm = env.get_vm(params["main_vm"])
+    tmp_file = "/tmp/vsock_file_%s" % utils_misc.generate_random_string(6)
+    session = vm.wait_for_login()
+    # TODO: temporary workaround for qemu bug, should be removed when fixed
+    session.cmd_output("setenforce 0")
+    nc_vsock_bin = compile_nc_vsock(test, vm, session)
+    vm_ip_addr = vm.wait_for_get_address(0)
+    port = utils_misc.find_free_port(5000, 6000, vm_ip_addr)
+    vsock_dev = params["vsocks"].split()[0]
+    guest_cid = vm.devices.get(vsock_dev).get_param("guest-cid")
+    nc_vsock_listen(nc_vsock_bin, port, session)
+    host_vsock_session = nc_vsock_connect(nc_vsock_bin, guest_cid, port)
+    connected_str = r"Connection from cid*"
+    send_data = "Hello world"
+    if session.read_until_last_line_matches(connected_str)[0] != 0:
+        clean()
+        test.fail("Connection failed.")
+    error_context.context('Input "Hello world" to vsock.', logging.info)
+    host_vsock_session.sendline(send_data)
+    if session.read_until_last_line_matches(send_data)[0] != 0:
```

Comment:
@vivianQizhu    The current code still have problem as my understanding,  `read_until_last_line_matches`  need a list  as parameter `patterns`.  and `send_data` here don't match it.  At last, it will only match the first character "H" in `Hello world`, same with `connected_str`.
refer:
```
 725     def match_patterns(cont, patterns):                                         
 736         for i in range(len(patterns)):                                          
 737             if not patterns[i]:                                                 
 738                 continue                                                        
 739             if re.search(patterns[i], cont):                                    
 740                 return i   
```
Assume the return value is `match, output`,  `match` means the matched index,  and i perfer to print the `output` when test failed.

Reply:
@yanan-fu You are right, we will get ` aexpect.exceptions.ExpectTimeoutError` if I pass the pattern correctly in list:
```
03:07:52 ERROR| aexpect.exceptions.ExpectTimeoutError: Timeout expired while looking for pattern 'Hello world'    (output: '[root@vm-74-183 home]# ')
```
Updated and tested, with selinux will get this:
```
03:38:34 ERROR| FAIL 1-Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.x86_64.io-github-autotest-qemu.vsock_test -> TestFail: Does not received expected content: Hello world, output of session: [root@vm-74-131 home]#
```